File: sdks/python/paycrest_sdk/registry.py
"""In-memory cache for aggregator public key + token catalogue."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable, Optional


@dataclass(frozen=True)
class SupportedToken:
    symbol: str
    contract_address: str
    decimals: int
    base_currency: str
    network: str


# Process-level static token registry. Populate with `register_token`
# at startup to skip the `/v2/tokens` round-trip for hot tokens
# (e.g. USDT on base). Lookup order in AggregatorRegistry.get_token:
# static registry -> cached `/tokens` fetch -> live `/tokens` fetch.
_STATIC_TOKENS: dict[str, SupportedToken] = {}


def _key(network: str, symbol: str) -> str:
    return f"{network.lower()}::{symbol.upper()}"


def register_token(token: SupportedToken) -> None:
    """Register a token statically so the gateway path resolves it
    without hitting `/v2/tokens`."""
    _STATIC_TOKENS[_key(token.network, token.symbol)] = token


def register_tokens(tokens: Iterable[SupportedToken]) -> None:
    for t in tokens:
        register_token(t)


def list_registered_tokens() -> list[SupportedToken]:
    return list(_STATIC_TOKENS.values())


def clear_registered_tokens() -> None:
    _STATIC_TOKENS.clear()


class AggregatorRegistry:
    def __init__(self, http_client, public_key_override: Optional[str] = None):
        self._http = http_client
        self._public_key: Optional[str] = None
        self._public_key_override = public_key_override
        self._tokens_by_network: dict[str, list[SupportedToken]] = {}

    def get_public_key(self) -> str:
        if self._public_key_override:
            return self._public_key_override
        if self._public_key:
            return self._public_key
        response = self._http.call("GET", "/pubkey")
        pem = response.get("data")
        if not isinstance(pem, str) or not pem:
            raise ValueError("Aggregator /pubkey returned no PEM data")
        self._public_key = pem
        return pem

    def get_tokens_for_network(self, network: str) -> list[SupportedToken]:
        slug = network.lower()
        if slug in self._tokens_by_network:
            return self._tokens_by_network[slug]
        response = self._http.call("GET", "/tokens", query={"network": slug})
        raw = response.get("data") or []
        tokens = [
            SupportedToken(
                symbol=row["symbol"],
                contract_address=row["contractAddress"],
                decimals=int(row["decimals"]),
                base_currency=row.get("baseCurrency", ""),
                network=row.get("network", slug),
            )
            for row in raw
        ]
        self._tokens_by_network[slug] = tokens
        return tokens

    def get_token(self, network: str, symbol: str) -> SupportedToken:
        # 1) Static registry (zero-RTT for hot tokens).
        static_hit = _STATIC_TOKENS.get(_key(network, symbol))
        if static_hit is not None:
            return static_hit

        # 2) Live fetch (with in-memory cache).
        tokens = self.get_tokens_for_network(network)
        want = symbol.upper()
        for token in tokens:
            if token.symbol.upper() == want:
                return token
        known = ", ".join(t.symbol for t in tokens) or "(none)"
        raise ValueError(
            f'Token "{symbol}" is not enabled on network "{network}". Known: {known}'
        )

    def preload(self, network: str) -> list[SupportedToken]:
        """Pre-warm the in-memory cache for a network."""
        return self.get_tokens_for_network(network)

File: sdks/go/sdk/registry.go
package sdk

import (
	"context"
	"fmt"
	"net/http"
	"strings"
	"sync"
)

// SupportedToken mirrors `GET /v2/tokens?network=<slug>` response rows.
type SupportedToken struct {
	Symbol          string `json:"symbol"`
	ContractAddress string `json:"contractAddress"`
	Decimals        int    `json:"decimals"`
	BaseCurrency    string `json:"baseCurrency"`
	Network         string `json:"network"`
}

// Process-level static token registry. Populate at startup with
// RegisterToken to skip the `/v2/tokens` round-trip on hot tokens.
var (
	staticTokensMu sync.RWMutex
	staticTokens   = map[string]SupportedToken{}
)

func staticTokenKey(network, symbol string) string {
	return strings.ToLower(network) + "::" + strings.ToUpper(symbol)
}

// RegisterToken adds a token to the static registry. Lookup order in
// the gateway path: static -> in-memory cache of /tokens -> live fetch.
func RegisterToken(token SupportedToken) {
	staticTokensMu.Lock()
	defer staticTokensMu.Unlock()
	staticTokens[staticTokenKey(token.Network, token.Symbol)] = token
}

// RegisterTokens bulk-registers a list of tokens.
func RegisterTokens(tokens []SupportedToken) {
	staticTokensMu.Lock()
	defer staticTokensMu.Unlock()
	for _, t := range tokens {
		staticTokens[staticTokenKey(t.Network, t.Symbol)] = t
	}
}

// ListRegisteredTokens returns a snapshot of the static registry.
func ListRegisteredTokens() []SupportedToken {
	staticTokensMu.RLock()
	defer staticTokensMu.RUnlock()
	out := make([]SupportedToken, 0, len(staticTokens))
	for _, t := range staticTokens {
		out = append(out, t)
	}
	return out
}

// ClearRegisteredTokens drops all entries. Test-only escape hatch.
func ClearRegisteredTokens() {
	staticTokensMu.Lock()
	defer staticTokensMu.Unlock()
	staticTokens = map[string]SupportedToken{}
}

func staticTokenLookup(network, symbol string) (SupportedToken, bool) {
	staticTokensMu.RLock()
	defer staticTokensMu.RUnlock()
	t, ok := staticTokens[staticTokenKey(network, symbol)]
	return t, ok
}

// aggregatorRegistry caches the aggregator's public key and per-network
// token catalog in-memory for the process lifetime.
type aggregatorRegistry struct {
	http             *httpClientConfig
	publicKeyMu      sync.RWMutex
	publicKey        string
	publicKeyOverride string
	tokensMu         sync.RWMutex
	tokensByNetwork  map[string][]SupportedToken
}

func newAggregatorRegistry(http *httpClientConfig, publicKeyOverride string) *aggregatorRegistry {
	return &aggregatorRegistry{
		http:              http,
		publicKeyOverride: publicKeyOverride,
		tokensByNetwork:   map[string][]SupportedToken{},
	}
}

func (r *aggregatorRegistry) getPublicKey(ctx context.Context) (string, error) {
	if r.publicKeyOverride != "" {
		return r.publicKeyOverride, nil
	}
	r.publicKeyMu.RLock()
	cached := r.publicKey
	r.publicKeyMu.RUnlock()
	if cached != "" {
		return cached, nil
	}

	var response APIResponse[string]
	if err := r.http.request(ctx, http.MethodGet, "/pubkey", nil, nil, &response); err != nil {
		return "", err
	}
	if response.Data == "" {
		return "", fmt.Errorf("aggregator /pubkey returned no PEM data")
	}

	r.publicKeyMu.Lock()
	r.publicKey = response.Data
	r.publicKeyMu.Unlock()
	return response.Data, nil
}

func (r *aggregatorRegistry) getTokensForNetwork(ctx context.Context, network string) ([]SupportedToken, error) {
	slug := strings.ToLower(network)
	r.tokensMu.RLock()
	cached, ok := r.tokensByNetwork[slug]
	r.tokensMu.RUnlock()
	if ok {
		return cached, nil
	}

	var response APIResponse[[]SupportedToken]
	if err := r.http.request(ctx, http.MethodGet, "/tokens", map[string]string{"network": slug}, nil, &response); err != nil {
		return nil, err
	}

	r.tokensMu.Lock()
	r.tokensByNetwork[slug] = response.Data
	r.tokensMu.Unlock()
	return response.Data, nil
}

func (r *aggregatorRegistry) getToken(ctx context.Context, network, symbol string) (SupportedToken, error) {
	// 1) Static registry — zero-RTT for hot tokens.
	if t, ok := staticTokenLookup(network, symbol); ok {
		return t, nil
	}

	// 2) Live fetch (with in-memory cache).
	tokens, err := r.getTokensForNetwork(ctx, network)
	if err != nil {
		return SupportedToken{}, err
	}
	want := strings.ToUpper(symbol)
	for _, t := range tokens {
		if strings.ToUpper(t.Symbol) == want {
			return t, nil
		}
	}
	known := make([]string, 0, len(tokens))
	for _, t := range tokens {
		known = append(known, t.Symbol)
	}
	return SupportedToken{}, fmt.Errorf("token %q is not enabled on network %q; known: %v", symbol, network, known)
}

// Preload warms the in-memory token cache for a network.
func (r *aggregatorRegistry) Preload(ctx context.Context, network string) ([]SupportedToken, error) {
	return r.getTokensForNetwork(ctx, network)
}
